### PR TITLE
sentencepiece bpe compatible tokenizer

### PR DESCRIPTION
**Pull request by eiz**
I believe this largely fixes the tokenization issues. The example mentioned in https://github.com/ggerganov/llama.cpp/issues/167 as well as my local tests (e.g. "accurately" should tokenize as `[7913, 2486]`) are fixed by it. Have not tested extensively though, especially with Unicode.

I saw some discussion around file format updates so just take this as an rfc, I just hacked something in

sorry if my coding style is not to your liking ;)

### CODE REVIEW

**Comment by slaren** [comment]
Breaks quantize.cpp currently, needs to update the tokenizer part to add the score.

**Comment by eiz** [comment]
doh, thanks for pointing that out, I've only been using fp16 =) will fix.

File: convert-pth-to-ggml.py
Comment (j-f1) [inline convert-pth-to-ggml.py:129]: Shouldn't this be added for all types of token, not just the regular text ones?

Comment (slaren) [inline convert-pth-to-ggml.py:129]: The score is only written for this type of token, but should be written for all.
```suggestion
        
        fout.write(struct.pack("f", tokenizer.get_score(i)))
```

**Review by j-f1 — COMMENTED** [review]
Neat! I think this is pretty close but the Unicode handling isn’t quite right. In particular I don’t believe the tokenizer should be UTF-8 aware, since LLaMA should be perfectly capable of handling invalid UTF-8 strings. It seems to operate on the byte level so I believe this PR as-is will prevent characters that are not in the token dataset from being tokenized. Unrecognized characters are currently represented using their UTF-8 bytes as separate tokens.

**Comment by eiz** [comment]
The handling of UTF-8 here is exactly the same as SentencePiece does. Multi-byte characters that don't form tokens will be output one byte at a time.

File: convert-pth-to-ggml.py
Comment (eiz) [reply convert-pth-to-ggml.py:129]: omg I broke this while fixing whitespace formatting up 😓

**Comment by j-f1** [comment]
That’s what happens when I do code review at 1am! Everything looks great now (but it’s still 1am so I am not going to approve it until tomorrow morning when I can take a proper look)

**Comment by ggerganov** [comment]
@eiz 
Before you merge this, add temporary notice at the top of the README to regen models (like this: https://github.com/ggerganov/llama.cpp/commit/8a01f565ff78cc6c0c5a9fa402787a2f179f2d78)

Also, let's start bumping the magic when the `ggml` models change:
https://github.com/ggerganov/llama.cpp/blob/master/convert-pth-to-ggml.py#L96

Or add a version `ggml` version number. Whichever you prefer.
Check during loading and if it is not correct - print message asking the user to regen the models.

P.S. Need a few more days before I start looking into details, so appreciate all the help from the collaborators so far

**Comment by slaren** [comment]
The tokenization look great, I couldn't find any differences with the original llama tokenizer.

**Comment by Ronsor** [comment]
@ggerganov I would suggest a version number. That allows for better error messages like `version unsupported` versus something like `invalid model file`.

**Comment by eiz** [comment]
"why not both?"
- changed file magic so existing unversioned files don't misparse (ggml -> ggmf "gg model file")
- now a version number in the header

**Review by Ronsor — APPROVED** [review]

File: quantize.cpp
Comment (j-f1) [inline quantize.cpp:83]: Suggestion: move the `format_version` to a shared header file of some sort, and then say `(unsupported version 2, expected 1)`

**Comment by eous** [comment]
Of all the tokenizers we've had this is by far my favorite. Nice work.

**Comment by ggerganov** [comment]
@eiz Apologies for the `convert-pth-to-ggml.py` - please resolve the conflicts and merge when you are ready

**Review by ggerganov — APPROVED** [review]

File: main.cpp
Comment (bakkot) [inline main.cpp:120]:
```suggestion
            fprintf(stderr, "%s: invalid model file '%s' (unsupported format version %" PRIu32 " - only version 1 is supported)\n",
```
Best to tell people not just that they have a wrong thing, but what they'd need to have instead.

(Also you might want to pull the supported version number out to a constant.)

Comment (bakkot) [reply main.cpp:120]: Oh lol I see `@j-f1` made exactly the same comment below, ignore me.

**Comment by bakkot** [comment]
Good news: I ran this using the scoring logic in https://github.com/ggerganov/llama.cpp/pull/270 and saw an improvement in perplexity for the 7B FP16 on wikitext-2/train from `10.4625` before this PR to `5.8149` after. That's a _huge_ improvement.

**Comment by vjeux** [comment]
I believe that there's an issue with this algorithm. It can only merge tokens if they are composite of existing tokens. Not only this, but the combination that works must be the highest scoring one.

I was curious to see in the llama vocabulary how many tokens would not tokenize to themselves and I got 1631 out of 32000.
https://gist.github.com/vjeux/5a466b4c47dc19ec9630f6fbf0cc3a1b

Note that I reimplemented the algorithm in this commit, so I may have made a mistake.

I'm trying to figure out right now if this is the same algorithm as in the sentencepiece project, in that case it's an issue with the original tokenizer. Or if it's an issue with this implementation only.

**Comment by vjeux** [comment]
@bobakfb asked ChatGPT to find the differences between this algo and SentencePiece one: https://github.com/google/sentencepiece/blob/master/src/bpe_model.cc

> Both of these algorithms appear to be implementations of the Byte-Pair Encoding (BPE) algorithm, which is often used for tokenizing text in natural language processing (NLP) tasks.
> 
> The key difference between the two is the post-processing step, specifically how they handle symbols or characters that were not matched to any token in the vocabulary.
> 
> Here are the notable differences:
> 
> Post-processing:
> In the first algorithm (which is implemented within the SentencePiece framework), if a symbol sequence doesn't exist in the vocabulary, the algorithm performs a recursive process of segmentation (resegmentation), trying to break down the sequence into smaller subwords or symbols that are present in the vocabulary. This means it attempts to encode unrecognized sequences to known pieces.
> 
> In the second algorithm (the llama_tokenizer), if a symbol sequence is not found in the vocabulary, it treats each individual character as a separate token and outputs their corresponding byte values (plus 3, as the first three positions seem reserved for special tokens). This approach treats unrecognized sequences as a series of individual characters.

It looks correct. There's a "resegment" step at the end of the sentencepiece algorithm ( https://github.com/google/sentencepiece/blob/master/src/bpe_model.cc#L175-L200 ) that isn't present in this implementation.

So we should add it as well.

**Pull request merged**
